src/openai/lib/azure.py

```diff
@@ -0,0 +1,321 @@
+        endpoint: str | None = None,
+        deployment: str | None = None,
+        api_key: str | None = None,
+        azure_ad_token: str | None = None,
+        azure_ad_token_provider: AzureADTokenProvider | None = None,
+        organization: str | None = None,
+        base_url: str | None = None,
+        timeout: float | Timeout | None | NotGiven = NOT_GIVEN,
+        max_retries: int = DEFAULT_MAX_RETRIES,
+        default_headers: Mapping[str, str] | None = None,
+        default_query: Mapping[str, object] | None = None,
+        http_client: httpx.Client | None = None,
+        _strict_response_validation: bool = False,
+    ) -> None:
+        if api_key is None:
+            api_key = os.environ.get("AZURE_OPENAI_API_KEY")
+        if api_key is None and azure_ad_token is None and azure_ad_token_provider is None:
+            raise OpenAIError(
+                "The api_key client option must be set either by passing api_key to the client or by setting the AZURE_OPENAI_API_KEY environment variable; If you're using Azure AD you should pass either the `azure_ad_token` or the `azure_ad_token_provider` argument."
+            )
+
+        if api_version is None:  # pyright: ignore[reportUnnecessaryComparison]
+            raise ValueError("Expected `api_version` to be given")
+
+        if default_query is None:
+            default_query = {"api-version": api_version}
+        else:
+            default_query = {"api-version": api_version, **default_query}
+
+        if base_url is None:
+            if endpoint is None:
+                raise ValueError("If base_url is not given, then endpoint must be given")
+
+            if deployment is not None:
+                base_url = f"{endpoint}/openai/deployments/{deployment}"
+            else:
+                base_url = f"{endpoint}/openai"
+
+        if api_key is None:
+            # define a sentinel value to avoid any typing issues
+            api_key = API_KEY_SENTINEL
+
+        super().__init__(
+            api_key=api_key,
+            organization=organization,
+            base_url=base_url,
+            timeout=timeout,
+            max_retries=max_retries,
+            default_headers=default_headers,
+            default_query=default_query,
+            http_client=http_client,
+            _strict_response_validation=_strict_response_validation,
+        )
+        self._azure_ad_token = azure_ad_token
+        self._azure_ad_token_provier = azure_ad_token_provider
+
+    def _get_azure_ad_token(self) -> str | None:
+        if self._azure_ad_token is not None:
+            return self._azure_ad_token
+
+        provider = self._azure_ad_token_provier
+        if provider is not None:
+            token = provider()
+            if not token or not isinstance(token, str):  # pyright: ignore[reportUnnecessaryIsInstance]
+                raise ValueError(
+                    f"Expected `get_ad_token` argument to return a string but it returned {token}",
```

Comment:
I think the error message should say `azure_ad_token_provider` here and in the async client

Reply:
Oops thanks for catching, thought I'd updated it!